src/server/pps/server/pipeline_state_driver.go

```diff
@@ -0,0 +1,272 @@
+	pipelineWatcher, err := sd.pipelines.ReadOnly(ctx).Watch()
+	if err != nil {
+		return nil, nil, errors.EnsureStack(err)
+	}
+	return pipelineWatcher.Watch(), pipelineWatcher.Close, nil
+}
+
+func (sd *stateDriver) ListPipelineInfo(ctx context.Context, f func(*pps.PipelineInfo) error) error {
+	return ppsutil.ListPipelineInfo(ctx, sd.pipelines, nil, 0, f)
+}
+
+func (sd *stateDriver) GetPipelineInfo(ctx context.Context, name string, version int) (*pps.PipelineInfo, error) {
+	var pipelineInfo pps.PipelineInfo
+	if err := sd.pipelines.ReadOnly(ctx).GetUniqueByIndex(
+		ppsdb.PipelinesVersionIndex,
+		ppsdb.VersionKey(name, uint64(version)),
+		&pipelineInfo); err != nil {
+		return nil, errors.Wrapf(err, "couldn't retrieve pipeline information")
+	}
+	return &pipelineInfo, nil
+}
+
+func (sd *stateDriver) tryLoadLatestPipelineInfo(ctx context.Context, pipeline string) (*pps.PipelineInfo, error) {
+	pi := &pps.PipelineInfo{}
+	errCnt := 0
+	err := backoff.RetryNotify(func() error {
+		return sd.loadLatestPipelineInfo(ctx, pipeline, pi)
+	}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+		errCnt++
+		// Don't put the pipeline in a failing state if we're in the middle
+		// of activating auth, retry in a bit
+		if (auth.IsErrNotAuthorized(err) || auth.IsErrNotSignedIn(err)) && errCnt <= maxErrCount {
+			log.Warnf("PPS master: could not retrieve pipelineInfo for pipeline %q: %v; retrying in %v",
+				pipeline, err, d)
+			return nil
+		}
+		return stepError{
+			error: errors.Wrapf(err, "could not load pipelineInfo for pipeline %q", pipeline),
+			retry: false,
+		}
+	})
+	return pi, err
+}
+
+func (sd *stateDriver) loadLatestPipelineInfo(ctx context.Context, pipeline string, message *pps.PipelineInfo) error {
+	specCommit, err := ppsutil.FindPipelineSpecCommit(ctx, sd.pfsApi, *sd.txEnv, pipeline)
+	if err != nil {
+		return errors.Wrapf(err, "could not find spec commit for pipeline %q", pipeline)
+	}
+	// TODO: which context should be passed here? master's ctx?
+	if err := sd.pipelines.ReadOnly(ctx).Get(specCommit, message); err != nil {
+		return errors.Wrapf(err, "could not retrieve pipeline info for %q", pipeline)
+	}
+	return nil
+}
+
+type mockStateDriver struct {
+	specCommits map[string]string              // maps spec commit IDs to pipeline names
+	pipelines   map[string]*pps.PipelineInfo   // maps pipeline names to PipelineInfos
+	states      map[string][]pps.PipelineState // tracks all of the
+	eChan       chan *watch.Event
+	doneEChan   chan struct{} // supports closing eChan
+}
+
+func newMockStateDriver() *mockStateDriver {
+	msd := &mockStateDriver{}
+	msd.reset()
+	msd.eChan = make(chan *watch.Event, 1)
+	msd.doneEChan = make(chan struct{}, 5)
+	return msd
+}
+
+func (msd *mockStateDriver) SetState(ctx context.Context, specCommit *pfs.Commit, state pps.PipelineState, reason string) error {
+	if pipeline, ok := msd.specCommits[specCommit.ID]; ok {
+		if pi, ok := msd.pipelines[pipeline]; ok {
```

Comment:
can't decide if it's important to capture, but strictly speaking `SetPipelineState` will only modify the version of the pipeline at that spec commit, which may no longer be current. This is important for concurrent updates, so I expect something like 

- send pipeline event for a failed pipeline while blocking `getRC` from completing
- update pipeline externally
- unblock `getRC` to let the old step continue with a stale pipeline info

to behave incorrectly in this model, with the pipeline staying in failure instead of going into starting etc

even if we don't want to capture that, this model feels a little weird - you could just be using the repo name from the spec commit, right? Or is the idea to test that the spec commit ID made it into the map properly?

Reply:
yeah the model here should be inverted so that `pipelines` -> `specCommit` -> `pipelineInfo`